src/core/src/op/read_value.cpp

```diff
@@ -50,36 +55,84 @@ op::v6::ReadValue::ReadValue(const Output<Node>& init_value, const shared_ptr<Va
-                          "Variable shape and output shape are inconsistent.");
-    m_variable->update(var_info);
-    set_output_type(0, arg_t, input_shape);
+    const auto& variable_info = m_variable->get_info();
+    const auto& variable_type = variable_info.data_type;
+    const auto& variable_shape = variable_info.data_shape;
+
+    // If no inputs provided, it means this ReadValue doesn't have initial subgraph. This is valid.
+    if (get_input_size() > 0) {
+        const auto& initial_type = get_input_element_type(0);
+        const auto& initial_shape = get_input_partial_shape(0);
+
+        // Variable's shape/type determine a permissible range of values for shape/type inferred from initial_subgraph.
+        // If initial_subgraph is set, then we need to check that shape/type inferred from initial_subgraph
+        // is within the permissible range.
+
+        const auto compatible_type = variable_type.is_dynamic() || initial_type == variable_type;
+        const auto compatible_shape = variable_shape.relaxes(initial_shape);
+
+        OPENVINO_ASSERT(compatible_shape,
+                        "The shape specified in the Variable has to extend (relax) the shape "
+                        "inferred from the initializing subgraph.",
+                        " Variable shape: ",
+                        variable_shape,
+                        " Initialization shape: ",
+                        initial_shape);
+        OPENVINO_ASSERT(compatible_type,
+                        "The type specified in the Variable has to extend (relax) the type "
+                        "inferred from the initializing subgraph.",
+                        " Variable type: ",
+                        variable_type,
+                        " Initialization type: ",
+                        initial_type);
+    }
+
+    set_output_type(0, variable_type, variable_shape);
 }
 
 shared_ptr<Node> op::v6::ReadValue::clone_with_new_inputs(const OutputVector& new_args) const {
     OV_OP_SCOPE(v6_ReadValue_clone_with_new_inputs);
     check_new_args_count(this, new_args);
-    return make_shared<ReadValue>(new_args.at(0), m_variable);
+
+    switch (new_args.size()) {
+    case 0:
+        return make_shared<ReadValue>(m_variable);
+    case 1:
+        return make_shared<ReadValue>(new_args[0], m_variable);
+    default:
+        OPENVINO_THROW("Unable to clone ReadValue ",
+                       this->get_friendly_name(),
+                       " Incorrect number of inputs. Expected: 0 or 1. Actual: ",
+                       new_args.size());
+    }
 }
 
 bool op::v6::ReadValue::visit_attributes(AttributeVisitor& visitor) {
     OV_OP_SCOPE(v6_ReadValue_visit_attributes);
     visitor.on_attribute("variable_id", m_variable);
+
+    auto variable_info = m_variable->get_info();
+    visitor.on_attribute("variable_type", variable_info.data_type);
+    visitor.on_attribute("variable_shape", variable_info.data_shape);
```

Comment:
Can we have the following interfaces for ReadValue? 
`get_variable_type` and `get_variable_shape`? Then plugin can retrieve them easily.

Reply:
@ilya-lavrenov @ilyachur @slyalin @itikhono We need to clarify a problem for stateful implementation & optimization in plugin -
1. The variable has a data type. 
2. The ngraph node ReadValue has an output data type. 
 
What's their relationship of these types? Do we expect them to be the same type? If not, which is the data type a user is assuming? Which data type should the plugin assume it should return this type as it is asked by user?